core/lib/via_btc_client/src/client/mod.rs

```diff
@@ -91,104 +96,212 @@ impl BitcoinOps for BitcoinClient {
+        self.rpc.get_block_by_height(block_height).await
+    }
+
+    #[instrument(skip(self), target = "bitcoin_client")]
     async fn get_transaction(&self, txid: &Txid) -> BitcoinClientResult<Transaction> {
+        debug!("Getting transaction");
         self.rpc.get_transaction(txid).await
     }
 
-    fn get_network(&self) -> Network {
-        self.network
+    #[instrument(skip(self), target = "bitcoin_client")]
+    async fn fetch_block_by_hash(&self, block_hash: &BlockHash) -> BitcoinClientResult<Block> {
+        debug!("Fetching block by hash");
+        self.rpc.get_block_by_hash(block_hash).await
     }
 }
 
 #[cfg(test)]
 mod tests {
-    use bitcoin::Network;
+    use std::str::FromStr;
+
+    use bitcoin::{absolute::LockTime, hashes::Hash, transaction::Version, Amount, Wtxid};
+    use bitcoincore_rpc::{
+        bitcoincore_rpc_json::GetBlockchainInfoResult,
+        json::{EstimateSmartFeeResult, GetRawTransactionResult},
+    };
+    use mockall::{mock, predicate::*};
 
     use super::*;
-    use crate::{regtest::BitcoinRegtest, traits::BitcoinOps};
+    use crate::types::BitcoinRpcResult;
+
+    mock! {
+        BitcoinRpc {}
+        #[async_trait]
+        impl BitcoinRpc for BitcoinRpc {
+            async fn get_balance(&self, address: &Address) -> BitcoinClientResult<u64>;
+            async fn send_raw_transaction(&self, tx_hex: &str) -> BitcoinClientResult<Txid>;
+            async fn list_unspent(&self, address: &Address) -> BitcoinClientResult<Vec<OutPoint>>;
+            async fn get_transaction(&self, txid: &Txid) -> BitcoinClientResult<Transaction>;
+            async fn get_block_count(&self) -> BitcoinClientResult<u64>;
+            async fn get_block_by_height(&self, block_height: u128) -> BitcoinClientResult<Block>;
+            async fn get_block_by_hash(&self, block_hash: &BlockHash) -> BitcoinClientResult<Block>;
+            async fn get_best_block_hash(&self) -> BitcoinClientResult<BlockHash>;
+            async fn get_raw_transaction_info(&self, txid: &Txid) -> BitcoinClientResult<GetRawTransactionResult>;
+            async fn estimate_smart_fee(&self, conf_target: u16, estimate_mode: Option<EstimateMode>) -> BitcoinClientResult<EstimateSmartFeeResult>;
+            async fn get_blockchain_info(&self) -> BitcoinRpcResult<GetBlockchainInfoResult>;
+        }
+    }
 
     #[tokio::test]
-    async fn test_new() {
-        let context = BitcoinRegtest::new().expect("Failed to create BitcoinRegtest");
-        let client = BitcoinClient::new(&context.get_url(), Network::Regtest)
-            .await
-            .expect("Failed to create BitcoinClient");
-
-        assert_eq!(client.network, Network::Regtest);
+    async fn test_get_balance() {
+        let mut mock_rpc = MockBitcoinRpc::new();
+        mock_rpc.expect_get_balance().return_once(|_| Ok(1000000));
+
+        let client = BitcoinClient {
+            rpc: Box::new(mock_rpc),
+            network: Network::Bitcoin,
+        };
```

Comment:
Maybe consider extracting this into a separate function to avoid repetitions. Just a suggestion.